lightning/src/ln/msgs.rs

```diff
@@ -306,6 +306,58 @@ impl NetAddress {
 	}
 }
 
+/// A "set" of addresses which enforces that there can be only up to one of each net address type.
+pub struct NetAddressSet {
+	v4: Option<NetAddress>,
+	v6: Option<NetAddress>,
+	onion2: Option<NetAddress>,
+	onion3: Option<NetAddress>,
+}
+impl NetAddressSet {
+	/// Creates a new, empty, NetAddressSet
+	pub fn new() -> Self {
+		NetAddressSet { v4: None, v6: None, onion2: None, onion3: None }
+	}
+
+	/// Sets the IPv4 socket address in this set, overwriting any previous IPv4 socket addresses
+	/// (if any).
+	pub fn set_v4(&mut self, addr: [u8; 4], port: u16) {
+		self.v4 = Some(NetAddress::IPv4 { addr, port });
+	}
```

Comment:
Rather than have each mutator take the components of the type as a parameter, could you have each take the type itself (e.g., `NetAddress::IPv4`) as the parameter instead?

Or, better yet, just make `NetAddressSet` a struct with public fields and no mutators. Then the user can also clear fields by assigning them `None`. This would also remove a lot of unnecessary code. The `Option` types enforce the constraints already.

Reply:
Hmm, I dont think there is a way to specify a variant of an enum as a method parameter, we'd have to split each variant of NetAddress into its own struct and make the NetAddress variants just hold the outer struct. I'm torn on if that is cleaner or not, so I'll defer to you.

Reply:
Ah, yeah. I didn't realize they were inlined, but yes pulling them out would be required.

I think it's cleaner especially since you wouldn't need to have mutators to enforce that each field holds the correct variant. The fact that each field is an option of the specific type makes it impossible for them to be set incorrectly. Less code is better IMHO.

Then `into_vec` could just wrap them in the appropriate enum variant.

Reply:
Hmm, so I started refactoring everything to just convert any Vec<NetAddress> to Option<v4>, Option<v6>, Option<onionv2>, ... and then realized we may at some point not even enforce that invariant - lnd is too lazy to enforce it and generates things that blatantly violate the spec already (and doesn't seem interested in fixing this at all), and I believe other implementations have adapted to accept their garbage too. CC https://github.com/lightningnetwork/lightning-rfc/pull/751

Reply:
I threw out all the crap limiting how many addresses of each type we accept. It never made a ton of sense and lnd has always completely ignored the requirements there. cc https://github.com/lightningnetwork/lightning-rfc/pull/751